### ISSUE TITLE
pixelDensity() not working in exported Windows applications

### ISSUE THREAD

**Issue by andrewtaylorjohnson** [processing/processing#5414]
## Description

When exporting a sketch to an application, displayDensity() no longer returns the correct value. This is happening on my high DPI monitor using Windows 10. Also, hardcoding the pixel density to 2, pixelDensity(2), has no effect whatsoever. This is not an issue when I run the sketch from the Processing editor.

## Expected Behavior

As I am using a high DPI monitor, I expect the displayDensity() function to return a value of 2. I also expect pixelDensity(2) to set the pixel density to 2.

## Current Behavior

displayDensity() returns a value of 1, and pixelDensity(2) has no effect.

## Steps to Reproduce

1. Within the setup function, use the following code snippet: pixelDensity(2);
2. Export the application and run.

## Your Environment

 

* Processing version: 3.3.6
* Operating System and OS version: Windows 10

**Comment by benfry** [comment]
Sounds like the OS isn't recognizing your display as high DPI.

**Comment by andrewtaylorjohnson** [comment]
Is it not strange that it would recognize it as high DPI from the editor but not from the build? I have other software (Photoshop, Maya, etc.) that recognized that it is and compensates just fine.

**Comment by benfry** [comment]
Are you checking the box to embed Java in the exported application?

**Comment by andrewtaylorjohnson** [comment]
I am indeed.

**Comment by JakubValtar** [comment]
@andrewtaylorjohnson Sorry, for now the density detection works only on Mac, on Win/Linux only when run from the editor. You can manually tell the sketch which screen density should be allowed:
- `--density=1` makes sure sketch always runs at 1x density,
- `--density=2` plus a call to `pixelDensity(2)` will make the sketch run at 2x density.

```processing
// add this code to the end of your sketch and change <sketch_name> to the name of your sketch
public static void main(String[] args) {
  int density = 2;
  String sketchName = <sketch_name>.class.getCanonicalName();
  String[] sketchArgs = {"--density=" + density, sketchName};
  PApplet.main(concat(sketchArgs, args));
}
```

@benfry we still don't have display detection in sketch because we would need to bundle the JNA, on Win/Linux it is detected in the editor and then sent to the sketch as a command line arg

**Comment by quinton-ashley** [comment]
This doesn't work for me.  The sketch only runs in the top left corner.

**Comment by JakubValtar** [comment]
@quinton-ashley I'm still waiting for your response in #5479. This is a different issue than you are having.

**Comment by quinton-ashley** [comment]
Oh okay.  Yeah when I used this code it made it fullscreen but I only saw the bottom left quarter of my program.

**Comment by philipp-lehmann** [comment]
Hi there

I had the same problem, but thanks to this (https://superuser.com/a/1207925) I could solve it.
This is how it looks on my win 10 system.

![high-dpi](https://user-images.githubusercontent.com/4519837/118241495-812c6400-b49c-11eb-9313-f3c72c6ede2f.jpg)

**Comment by benfry** [comment]
Should be resolved in 4.0 beta 6 when it's available: https://github.com/processing/processing4/releases
